### ISSUE TITLE
Implementing WAIC

### ISSUE THREAD

**Issue by aloctavodia** [pymc-devs/pymc#888]
I think would be nice to have the WAIC (Watanabe-Akaike information criterion) implemented on PyMC3. In [this paper](http://www.stat.columbia.edu/~gelman/research/unpublished/waic_stan.pdf) Gelman discuss WAIC, leave-one-out cross-validation and K-fold cross-validation in the context of Stan. I think the following code is a correct implementation of waic. I would appreciate if someone could check it.

``` python
def waic(model, trace):
    """
    Calculate the widely available information criterion
    """
    obs = model.observed_RVs[0]
    log_py = [obs.logp_elemwise(pt) for pt in trace]
    lppd =  np.sum(np.log(np.mean(np.exp(log_py), axis=0)))

    p_waic = np.sum(np.var(log_py, axis=0))

    return -2 * lppd + 2 * p_waic
```

What do you think, in general terms, of these types of criteria for model selection?

**Comment by twiecki** [comment]
Thanks @aloctavodia. WAIC support (as well as DIC) would be great for PyMC3. Your approach looks correct although I haven't checked the math in detail but trust you did that correctly. Want to do a PR?

An alternative idea would be to implement this as a method of the Model class so that one could do `model.waic(trace)`. On second thought, that might break the pattern where we get the model object from the context. So alternatively, you could do that one could do `with model: waic(trace)`

**Comment by aloctavodia** [comment]
I agree on the `with model: waic(trace)` alternative. I will check how DIC was implemented. I also need to write a general version for more than one observed_RVs and do a PR. I will try to do this next week.

**Comment by twiecki** [comment]
@aloctavodia Great. I think we don't have DIC yet.

**Comment by fonnesbeck** [comment]
Wait, we do have DIC as of #834 (corrected from #797)

**Comment by twiecki** [comment]
Oh, very cool.
